### PR TITLE
Test/core api

### PR DESCRIPTION
**Pull request by futufeld**
This change adds tests of the core API functionality. The only changes (besides added tests) to the project are:
- Changes to the .cabal file to enable test compilation.
- Refactoring of Spec.hs to separate tests with different concerns.
- Added the Show and Eq typeclasses to some types to enable HSpec functionality.

### CODE REVIEW

File: test/StateSpec.hs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+defAppState :: AppState
+defAppState = def
+
+scheduler' :: Scheduler -> Result
+scheduler' = scheduler defAppState
+
+scheduler :: AppState -> Scheduler -> Result
+scheduler aState m = (x, y ^. appState)
+    where (x, y) = ES.runEState aState . runExceptT $ m
```
Comment (rimmington) [inline test/StateSpec.hs]: `where (x, y) = ES.runEState aState $ runExceptT`
or alternatively
`scheduler aState = fmap (^. appState) . ES.runEState aState . runExceptT`

File: test/StateSpec.hs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+dependentJob :: Name -> [Name] -> JobSpec
+dependentJob job deps = dependsOn .~ deps $ isolatedJob job
+
+shouldHave :: Result -> (Result -> Either String Bool) -> Expectation
+shouldHave result f = result `shouldSatisfy` (either (const False) (id)) . f
+
+getJob :: Name -> Result -> Either String JobStatus
+getJob jName result = maybe (Left "No job found.") Right job
+    where job = (snd result) ^. appState ^. jobs ^. at jName
```
Comment (rimmington) [inline test/StateSpec.hs]: Prefer pattern matching over `snd`.

File: test/StateSpec.hs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+noJob :: Name -> Result -> Either String Bool
+noJob jName = validator
+    where validator result
+            | (isLeft . fst) result = Left "Process returned error."
+            | otherwise = Right $ isLeft (getJob jName result)
```
Comment (rimmington) [inline test/StateSpec.hs]: You can inline `validator` and (probably) use pattern guards here. Also, you might want to preserve the contents of the error (whatever that is) for debugging.

File: test/StateSpec.hs [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+getRevDep :: Name -> Result -> Either String [Name]
+getRevDep jName result = maybe (Left "No reverse dependency found.") Right revDep
+    where revDep = (snd result) ^. appState ^. revDeps ^. at jName
```
Comment (rimmington) [inline test/StateSpec.hs]: Again, prefer pattern matches over `snd`.

File: test/StateSpec.hs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    where validator result
+            | (isLeft . fst) result = Left "Process returned error."
+            | otherwise = Right $ isLeft (getRevDep jName result)
+
+jobInState :: Name -> JobState -> Result -> Either String Bool
+jobInState jName jState = validator
+    where validator result
+            | (isLeft . fst) result = Left "Process returned error."
+            | otherwise = fmap inState $ getJob jName result
+          inState = (== jState) . (flip (^.) jobState)
```
Comment (rimmington) [inline test/StateSpec.hs]: `inState = (== jState) . view jobState`

File: test/StateSpec.hs [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+jobWithRevDep :: Name -> [Name] -> Result -> Either String Bool
+jobWithRevDep jName jRevDeps = validator
+    where validator result
+            | (isLeft . fst) result = Left "Process returned error."
+            | otherwise = fmap (== jRevDeps) $ getRevDep jName result
+
+producedError :: JobError -> Result -> Either String Bool
+producedError jError = \m -> case fst m of
+    Left x -> Right $ x == jError
+    Right x -> Left "Job did not produce error."
```
Comment (rimmington) [inline test/StateSpec.hs]: Unnecessary lambda, and you can pattern match on `m`.

File: test/StateSpec.hs [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            let result = createJob' $ isolatedJob "job"
+            in (scheduler' result) `shouldHave` (jobInState "job" $ Queued LocalQueue)
+
+    context "when provided with dependency in Finished state" $
+        it "should be placed in the Queued state" $
+            let result = do
+                    createJob' $ isolatedJob "dep"
+                    lift $ ES.setJobState "dep" Finished
+                    createJob' $ dependentJob "job" ["dep"]
+            in (scheduler' result) `shouldHave` (jobInState "job" $ Queued LocalQueue)
```
Comment (rimmington) [inline test/StateSpec.hs]: Is there a reason the let blocks are indented so far?

**Comment by rimmington** [comment]
If you want, you can run hlint over the spec files. I'm going to run it over the entire project later anyway.

File: test/StateSpec.hs (hunk 7 above)
Comment (futufeld) [inline test/StateSpec.hs]: Using this case as an example, `createJob'` has to exceed `result` in terms of indentation for the file to parse. It is visually off-putting, though. Are there any alternatives, besides reducing the offset (fewer spaces) in these cases?

**Comment by futufeld** [comment]
I recall hlint but just didn't get around to running it (due to forgetfulness, rather than neglect). I'll work on making it a habit, and running it after making the changes we discussed will be a good opportunity to get into that workflow.

File: test/StateSpec.hs (hunk 1 above)
Comment (futufeld) [inline test/StateSpec.hs]: Somehow I completely missed the functor instance for tuples. That's handy.

**Comment by futufeld** [comment]
Thanks for the feedback. They're good suggestions that I'd like to see in the code, so I'll update the tests.

**Pull request merged**